src/coreclr/System.Private.CoreLib/src/System/Threading/Interlocked.CoreCLR.cs

```diff
@@ -48,27 +48,60 @@ public static partial class Interlocked
         [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int Exchange(ref int location1, int value)
+        {
+#if TARGET_X86 || TARGET_AMD64 || TARGET_ARM64 || TARGET_RISCV64
+            return Exchange(ref location1, value); // Must expand intrinsic
+#else
+            if (Unsafe.IsNullRef(ref location1))
+                ThrowHelper.ThrowNullReferenceException();
+            return Exchange32(ref location1, value);
+#endif
+        }
+
         [MethodImpl(MethodImplOptions.InternalCall)]
-        public static extern int Exchange(ref int location1, int value);
+        private extern static int Exchange32(ref int location1, int value);
 
         /// <summary>Sets a 64-bit signed integer to a specified value and returns the original value, as an atomic operation.</summary>
         /// <param name="location1">The variable to set to the specified value.</param>
         /// <param name="value">The value to which the <paramref name="location1"/> parameter is set.</param>
         /// <returns>The original value of <paramref name="location1"/>.</returns>
         /// <exception cref="NullReferenceException">The address of location1 is a null pointer.</exception>
         [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static long Exchange(ref long location1, long value)
+        {
+#if TARGET_AMD64 || TARGET_ARM64 || TARGET_RISCV64
+            return Exchange(ref location1, value); // Must expand intrinsic
+#else
+            if (Unsafe.IsNullRef(ref location1))
+                ThrowHelper.ThrowNullReferenceException();
+            return Exchange64(ref location1, value);
+#endif
+        }
+
         [MethodImpl(MethodImplOptions.InternalCall)]
-        public static extern long Exchange(ref long location1, long value);
+        private extern static long Exchange64(ref long location1, long value);
 
         /// <summary>Sets an object to the specified value and returns a reference to the original object, as an atomic operation.</summary>
         /// <param name="location1">The variable to set to the specified value.</param>
         /// <param name="value">The value to which the <paramref name="location1"/> parameter is set.</param>
         /// <returns>The original value of <paramref name="location1"/>.</returns>
         /// <exception cref="NullReferenceException">The address of location1 is a null pointer.</exception>
         [Intrinsic]
-        [MethodImpl(MethodImplOptions.InternalCall)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         [return: NotNullIfNotNull(nameof(location1))]
-        public static extern object? Exchange([NotNullIfNotNull(nameof(value))] ref object? location1, object? value);
+        public static object? Exchange([NotNullIfNotNull(nameof(value))] ref object? location1, object? value)
+        {
+            if (Unsafe.IsNullRef(ref location1))
```

Comment:
The change moves the null check + throw from the C/C++ to C#. The JIT is smart enough to eliminate the null check in number of cases. I was not able to measure any actual perf improvement from that. The CAS operation dominates the perf and the extra null check is noise.